### PR TITLE
Multizone Config Update 

### PR DESCRIPTION
**Pull request by talbring**
## Proposed Changes
This is the first part of a series of PRs which completely refactor the output (and parts of the input).

In this one I introduce the sub-config file feature for multizone/multiphysics problems. More information on how it works can be found here:

https://su2code.github.io/docs/Multizone

Note that the automatic deduction of the correct interface type is *not* working at the moment. So the `MARKER_FLUID_INTERFACE` or `MARKER_CHT_INTERFACE` still has to be there at the moment ...

Furthermore this changes the way grid movement is specified in the config files. Examples can be found also using the link above or by simply looking at the `sliding_interface` test cases. 

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [x] My contribution is commented and consistent with SU2 style.
- [x] I have added a test case that demonstrates my contribution, if necessary.

### CODE REVIEW

File: Common/include/option_structure.hpp [hunk 1]
```diff
@@ -626,38 +626,48 @@ static const map<string, ENUM_CONDUCTIVITYMODEL_TURB> TurbConductivityModel_Map
...
-  EXTERNAL_ROTATION = 5,  /*!< \brief Arbitrary grid motion specified by external files at each time step with rigid rotation. */
-  AEROELASTIC = 6,    /*!< \brief Simulation with aeroelastic motion. */
-  MOVING_WALL = 7,    /*!< \brief Simulation with moving walls (translation/rotation). */
   ROTATING_FRAME = 8,    /*!< \brief Simulation in a rotating frame. */
   ELASTICITY = 9,    /*!< \brief Linear Elasticity. */
-  AEROELASTIC_RIGID_MOTION = 10, /*!< \brief Simulation with rotation and aeroelastic motion. */
   STEADY_TRANSLATION = 11,    /*!< \brief Simulation in a steadily translating frame. */
   GUST = 12, /*!< \brief Simulation on a static mesh with a gust. */
   MOVING_HTP = 13,    /*!< \brief Simulation with moving HTP (rotation). */
-  FLUID_STRUCTURE_STATIC = 14 /*!< \brief Fluid structure deformation with no grid velocity. */
```
Comment (WallyMaier) [inline Common/include/option_structure.hpp]: Should the identifiers be updated?

File: SU2_CFD/src/numerics_direct_mean.cpp [hunk 2]
```diff
@@ -5768,9 +5768,9 @@ void CSourceRotatingFrame_Flow::ComputeResidual(su2double *val_residual, su2doub
   
   /*--- Retrieve the angular velocity vector from config. ---*/
   
-  Omega[0] = config->GetRotation_Rate_X(config->GetiZone())/config->GetOmega_Ref();
-  Omega[1] = config->GetRotation_Rate_Y(config->GetiZone())/config->GetOmega_Ref();
-  Omega[2] = config->GetRotation_Rate_Z(config->GetiZone())/config->GetOmega_Ref();
+  for (iDim = 0; iDim < 3; iDim++){
+    Omega[iDim] = config->GetRotation_Rate()[iDim]/config->GetOmega_Ref();
```
Comment (WallyMaier) [inline SU2_CFD/src/numerics_direct_mean.cpp]: These changes are great, thanks!

File: Common/include/config_structure.inl [hunk 3]
```diff
@@ -937,71 +937,51 @@ inline void CConfig::SetMu_PolyCoeffND(su2double val_coeff, unsigned short val_i
 
 inline void CConfig::SetKt_PolyCoeffND(su2double val_coeff, unsigned short val_index) { KtPolyCoefficientsND[val_index] = val_coeff; }
 
-inline unsigned short CConfig::GetKind_GridMovement(unsigned short val_iZone) { return Kind_GridMovement[val_iZone]; }
+inline unsigned short CConfig::GetKind_GridMovement() { return Kind_GridMovement; }
```
Comment (economon) [inline Common/include/config_structure.inl]: Nice.. the multi config structure makes things like the grid motion specification much simpler 👍

File: Common/include/option_structure.hpp [hunk 4]
```diff
@@ -2059,11 +2069,13 @@ static const map<string, ENUM_VERIFICATION_SOLUTIONS> Verification_Solution_Map
 
 class COptionBase {
 private:
+  vector<string> value;
 public:
   COptionBase() {};
   virtual  ~COptionBase() = 0;
   //  virtual string SetValue(string) {SU2MPI::PrintAndFinalize("shouldn't be here"); return "";};
```
Comment (economon) [inline Common/include/option_structure.hpp]: can remove this outdated commented line

File: Common/src/config_structure.cpp [hunk 5]
```diff
@@ -2562,16 +2663,138 @@ bool CConfig::SetRunTime_Parsing(char case_filename[MAX_STRING_SIZE]) {
...
+      
+    }
+    
+    /*--- Check if subconfig files exist --- */
+    
+    if (nConfig_Files != 0){
+      for (unsigned short iConfig = 0; iConfig < nConfig_Files; iConfig++){
+        ifstream f(Config_Filenames[iConfig].c_str());
+        if (!f.good()){
+          SU2_MPI::Error("Config file " + Config_Filenames[iConfig] + " defined in CONFIG_FILES does not exist", CURRENT_FUNCTION);
```
Comment (economon) [inline Common/src/config_structure.cpp]: thanks for the error checks 👍

File: Common/src/geometry_structure.cpp [hunk 6]
```diff
@@ -13760,205 +13759,6 @@ void CPhysicalGeometry::SetMaxLength(CConfig* config) {
 
 }
 
-void CPhysicalGeometry::MatchInterface(CConfig *config) {
```
Comment (economon) [inline Common/src/geometry_structure.cpp]: thanks for cleaning out marker_interface everywhere (including the old MPI)

File: SU2_CFD/src/iteration_structure.cpp [hunk 7]
```diff
@@ -107,8 +110,36 @@ void CIteration::SetGrid_Movement(CGeometry ****geometry_container,
...
+
+      break;
+
+ 	/*--- Already initialized in the static mesh movement routine at driver level. ---*/ 
+    case STEADY_TRANSLATION: case ROTATING_FRAME:
+      break;
+
+  }
+  
+  if (config_container[val_iZone]->GetSurface_Movement(DEFORMING)){
```
Comment (economon) [inline SU2_CFD/src/iteration_structure.cpp]: One thing I have been meaning to change is to remove the calls to SetVolume_Deformation() in between each of the updates of the surface positions.. in the case that multiple sequential surface movements are present, they should be additive (they use an increment to the position) and we should be able to call SetVolume_Deformation() just once at the end of the sequence. Can you please give it a look and adjust if you agree?

File: SU2_CFD/src/solver_direct_mean.cpp [hunk 8]
```diff
@@ -2765,6 +2449,7 @@ void CEulerSolver::SetNondimensionalization(CConfig *config, unsigned short iMes
   Time_Ref          = Length_Ref/Velocity_Ref;                                     config->SetTime_Ref(Time_Ref);
   Omega_Ref         = Velocity_Ref/Length_Ref;                                     config->SetOmega_Ref(Omega_Ref);
   Force_Ref         = config->GetDensity_Ref()*Velocity_Ref*Velocity_Ref*Length_Ref*Length_Ref; config->SetForce_Ref(Force_Ref);
+  Heat_Flux_Ref     = Density_Ref*Velocity_Ref*Velocity_Ref*Velocity_Ref;           config->SetHeat_Flux_Ref(Heat_Flux_Ref);
```
Comment (economon) [inline SU2_CFD/src/solver_direct_mean.cpp]: thx, this has been missing 👍 . we should also use it in BC_Heat_Flux() to non-dim the input heat flux at the wall I think to be consistent (see incompressible version of that BC)

File: Common/include/config_structure.inl [hunk 9]
```diff
@@ -937,71 +937,51 @@ inline void CConfig::SetMu_PolyCoeffND(su2double val_coeff, unsigned short val_i
...
-inline void CConfig::SetMotion_Origin_X(unsigned short val_iZone, su2double val_origin) { Motion_Origin_X[val_iZone] = val_origin; }
-
-inline void CConfig::SetMotion_Origin_Y(unsigned short val_iZone, su2double val_origin) { Motion_Origin_Y[val_iZone] = val_origin; }
-
-inline void CConfig::SetMotion_Origin_Z(unsigned short val_iZone, su2double val_origin) { Motion_Origin_Z[val_iZone] = val_origin; }
-
-inline su2double CConfig::GetTranslation_Rate_X(unsigned short val_iZone) { return  Translation_Rate_X[val_iZone]; }
-
-inline su2double CConfig::GetTranslation_Rate_Y(unsigned short val_iZone) { return  Translation_Rate_Y[val_iZone]; }
+inline su2double* CConfig::GetMarkerMotion_Origin(unsigned short iMarkerMoving){return &MarkerMotion_Origin[3*iMarkerMoving];}
```
Comment (pcarruscag) [inline Common/include/config_structure.inl]: Can you return const pointer instead?

File: Common/src/fem_geometry_structure.cpp [hunk 10]
```diff
@@ -6929,4 +6863,66 @@ void CMeshFEM_DG::InitStaticMeshMovement(CConfig              *config,
...
+    for(unsigned short i=0; i<boundaries.size(); ++i) {
+      if( !boundaries[i].periodicBoundary ) {
+        
+        /* Check if for this boundary a motion has been specified. */
+        if (config->GetMarker_All_Moving(i) == YES) {
+          
+          /* Determine the prescribed translation velocity, rotation rate
+               and rotation center. */
+          const su2double Center[] = {config->GetMotion_Origin()[0],
+                                      config->GetMotion_Origin()[1],
```
Comment (pcarruscag) [inline Common/src/fem_geometry_structure.cpp]: If this is the general way in which these methods will be called, why not pass iDim as an argument and void returning pointers?

Comment (talbring) [reply Common/src/fem_geometry_structure.cpp]: Thanks for pointing that out. Thats something I wanted to change, thanks for reminding me :)

File: SU2_CFD/src/iteration_structure.cpp (hunk 7 above)
Comment (talbring) [reply SU2_CFD/src/iteration_structure.cpp]: Maybe we can deal with that in the new mesh solver implementation @rsanfer ?

Comment (economon) [reply SU2_CFD/src/iteration_structure.cpp]: Agreed. Let's wait on this change. I will make an issue instead.

**Comment by stephansmit** [comment]
Was playing around with the turbomachinery cases with the new multi-zone config. I like the new changes, makes it much clearer!

Just a remark, don't know if this is on purpose, but for the configuration parameter TURBOMACHINERY_KIND, you still need to supply them in the general configuration file for both zones (so in case of a two zone problem: TURBOMACHINERY_KIND = CENTRIFUGAL CENTRIFUGAL). Wouldnt it make more sense to have this per zone specified in the respective configuration file?

**Pull request merged**